### DISCUSSION TITLE
React way to render content outside "main" editor

### DISCUSSION THREAD

**Discussion by LoiKos** [ueberdosis/tiptap#2657] [General Questions] [answered]
Hi there,

First of all, thanks for open-sourcing this lib, DX is really great compare to some others text-editor lib.

I just started using this lib a couple of days ago, and i used it to build a comment Panel for my app (React). 
At this time i'm fine with the editor but i'm still searching how i could/should render "Comment" outside. 

If i understand correctly the docs and these :
https://github.com/ueberdosis/tiptap/issues/657 
https://github.com/ueberdosis/tiptap/discussions/241

I have 2 options : 

1.  Using another Editor with `editable:false` 
e.g 
```ts
interface CommentProps {
  content: JSONContent | string
}

function Comment(props: CommentProps) {
  const editor = useEditor({
    editable: false,
    extensions: [StarterKit],
  })

  React.useEffect(() => {
    editor?.commands.setContent(props.content)
  }, [props.content, editor])

  return <EditorContent editor={editor} />
}

export default Comment
```
2. Render HTML directly
```ts
interface CommentProps {
  content: string | JSONContent
}

function createMarkup(content: string | JSONContent) {
  return {
    __html:
      typeof content === 'string'
        ? content
        : generateHTML(content, [StarterKit]),
  }
}

function Comment(props: CommentProps) {
  return <div dangerouslySetInnerHTML={createMarkup(props.content)} />
}

export default Comment
```

First method seem easier to use because it should be fully compatible. but what about performances. Build an editor for every message doesn't bring too much overload ? 

Second method look ok but surely need to add some code to prevent XSS attacks on createMarkup i believe or does generateHTML handle that directly ?

Is these methods the recommanded way in Tiptap v2 to render content ? Is there anything i missed on this?

**Comment by philippkuehn** [comment, marked as answer]
Of course, this depends on how many editors are rendered and how complex they are. Basically, it is no problem to render several editors. At a certain point, it might make sense to use virtual list rendering.

Tiptap has a very strict schema. You can’t use any HTML element or attribute that is not defined in your schema. But there is an open issue/PR for links you have to keep in mind: https://github.com/ueberdosis/tiptap/pull/2169
